src/common/utils/use-waiting-message.ts

```diff
@@ -0,0 +1,48 @@
+import { useState } from 'react';
+
+// Keys are the seconds to wait before showing the message
+export type WaitingMessages = {
+  [key: string]: string;
+};
+
+function messageForSecondsPassed(waitingMessages: WaitingMessages, seconds: number) {
+  return waitingMessages[seconds.toString() as keyof typeof waitingMessages];
+}
+
+export function useWaitingMessage(
+  waitingMessages: WaitingMessages,
+  { waitingMessageInterval, maxCheckWaitingTime } = {
+    waitingMessageInterval: 1000,
+    maxCheckWaitingTime: 70000,
+  }
+): [string, () => void, () => void] {
+  let stopped = false;
+  let startUnlockTimeMs: number | undefined;
+  const [waitingMessage, setWaitingMessage] = useState(messageForSecondsPassed(waitingMessages, 0));
+
+  function startWaitingMessage() {
+    stopped = false;
+    startUnlockTimeMs = performance.now();
+    showMessageAtInterval();
+  }
+
+  function stopWaitingMessage() {
+    stopped = true;
+  }
+
+  function showMessageAtInterval() {
+    let waitingMessageInner;
+
+    setTimeout(() => {
```

Comment:
To avoid state being changed after a component has been destroyed and throwing an error, we need to cancel the timeout.

https://www.joshwcomeau.com/snippets/react-hooks/use-timeout/